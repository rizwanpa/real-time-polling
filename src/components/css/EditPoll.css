input, label {
    display:block;
}

.title {
    height: 24px;
    font-size: 14px;
    position: relative;
    display: -ms-inline-flexbox;
    display: inline-flex;
    -ms-flex-align: center;
    align-items: center;
    height: 32px;
    color: rgba(0, 0, 0, 0.85);
    font-size: 14px
}
.title::before {
    display: inline-block;
    margin-right: 4px;
    color: #ff4d4f;
    font-size: 14px;
    font-family: SimSun, sans-serif;
    line-height: 1;
    content: '*';
}
.input-title{
    width: 100%;
    padding: 12px 20px;
    margin-top: 10px;
    box-sizing: border-box;
    -webkit-box-sizing: border-box;
    box-sizing: border-box;
    font-variant: tabular-nums;
    list-style: none;
    -webkit-font-feature-settings: 'tnum';
    font-feature-settings: 'tnum', "tnum";
    position: relative;
    display: inline-block;
    min-width: 0;
    color: rgba(0, 0, 0, 0.85);
    font-size: 14px;
    line-height: 1.5715;
    background-color: #fff;
    background-image: none;
    border: 1px solid #d9d9d9;
    border-radius: 2px;
    -webkit-transition: all 0.3s;
    transition: all 0.3s;
}
.desc-title{
    height: 24px;
    font-size: 14px;
    padding-top: 10px;
    position: relative;
    -ms-flex-align: center;
    align-items: center;
    height: 32px;
    color: rgba(0, 0, 0, 0.85);
    font-size: 14px
}
.text-area-input {
    max-width: 100%;
    height: auto;
    min-height: 32px;
    line-height: 1.5715;
    vertical-align: bottom;
    -webkit-transition: all 0.3s, height 0s;
    transition: all 0.3s, height 0s;
}
.questions{
    margin: 10px;
    width: 90%;
}
.input-question{
    padding: 5px 10px
}
.options{
    margin: 10px;
    width: 60%;
}
.margin10{
    margin: 10px;
}
.btn-update{
    align-items: center;
    justify-content: center;
    text-align: center;
    padding: 10px;
}