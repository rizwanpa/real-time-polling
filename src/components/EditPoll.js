import React, { Component } from "react";
import { connect } from "react-redux";
import moment from "moment";
import { getPoll, editPoll, deletePollOption, deletePollQuestion } from "./../actions";
import {
  Divider,
  Form,
  Input,
  Button,
  Radio,
  Space,
  DatePicker,
  Row,
  Col,
  message,
  Modal
} from "antd";
import "./css/EditPoll.css";
import { MinusCircleOutlined, PlusOutlined } from "@ant-design/icons";

class EditPoll extends Component {
  constructor(props) {
    super(props);
    this.state = {
      editPoll: []
    };
  }
  componentDidMount() {
    console.log(
      "inside edit---componentDidMount",
      this.props,
      this.props.match.params.pollId
    );
    if (this.props.match.params.pollId) {
      this.props.getPollAction(this.props.match.params.pollId);
    }
  }
  componentWillReceiveProps(nextProps) {
    console.log("###############receiveProps==>", nextProps);
    this.setState({
      editPoll: nextProps.editPoll
    });
    if (
      !nextProps.editDetails.errorCode &&
      nextProps.editDetails.errorCode == 0 &&
      nextProps.editPoll.length == 0
    ) {
      message.success(nextProps.editDetails.data);
      this.props.history.push(`/polls`);
    } else {
      //message.error('OPPs something went wrong!');
    }
  }
  componentDidUpdate() {
    console.log("inside edit---componentDidUpdate", this.props, this.state);
    if (this.props.match.params.uuid) {
    }
  }
  onChangeHandledTitle = e => {
    console.log("change tilte==>", e.target.name, e.target.value, e.target);
    let updateval = { [e.target.name]: e.target.value };
    this.setState(prevState => {
      let editPoll = Object.assign({}, prevState.editPoll); // creating copy of state variable editPoll
      editPoll = {
        ...editPoll,
        ...updateval
      }; // update the name property, assign a new value
      return { editPoll }; // return new object jasper object
    });
  };
  onChangeHandledQuestion = (value, index, question_id) => {
    this.setState(prevState => {
      let editPoll = Object.assign({}, prevState.editPoll); // creating copy of state variable editPoll
      editPoll.questions[index].question = value;
      editPoll = {
        ...editPoll
      }; // update the name property, assign a new value
      return { editPoll }; // return new object jasper object
    });
  };
  onChangeHandledOption = (value, questionIndex, optionIndex) => {
    this.setState(prevState => {
      let editPoll = Object.assign({}, prevState.editPoll); // creating copy of state variable editPoll
      editPoll.questions[questionIndex].options[optionIndex].option = value;
      editPoll = {
        ...editPoll
      }; // update the name property, assign a new value
      return { editPoll }; // return new object jasper object
    });
  };
  onChangeType = (value, questionIndex) => {
    console.log("onChageType===>", value, questionIndex);
    this.setState(prevState => {
      let editPoll = Object.assign({}, prevState.editPoll);
      editPoll.questions[questionIndex].type = value;
      editPoll = {
        ...editPoll
      };
      return { editPoll };
    });
  };
  addQuestion = () => {
    this.setState(prevState => {
      let editPoll = Object.assign({}, prevState.editPoll);
      editPoll.questions.push({
        question: "",
        type: 0,
        options: []
      });
      editPoll = {
        ...editPoll
      };
      return { editPoll };
    });
  };
  addOption = questionIndex => {
    this.setState(prevState => {
      let editPoll = Object.assign({}, prevState.editPoll);
      if (!editPoll.questions[questionIndex].options) {
        editPoll.questions[questionIndex]["options"] = [];
      }
      editPoll.questions[questionIndex].options.push({
        option: ""
      });
      editPoll = {
        ...editPoll
      };
      return { editPoll };
    });
  };
  updatePoll = () => {
    this.props.editPollAction(this.state.editPoll);
  };
  onChangeStatus = value => {
    this.setState(prevState => {
      let editPoll = Object.assign({}, prevState.editPoll);
      editPoll.status = value;
      editPoll = {
        ...editPoll
      };
      return { editPoll };
    });
  };
  removeOption = (questionIndex, optionIndex) =>{
    this.setState(prevState => {
      let editPoll = Object.assign({}, prevState.editPoll);
      console.log('editPoll--->',editPoll.questions[questionIndex].options[optionIndex].id);
      let optionId = editPoll.questions[questionIndex].options[optionIndex].id;
      if(optionId){
        this.props.deletePollOptionAction(optionId)
      }
      editPoll.questions[questionIndex].options.splice(optionIndex, 1)
      return { editPoll };
    });
  }
  removeQuestion = (questionIndex) =>{
    this.setState(prevState => {
      let editPoll = Object.assign({}, prevState.editPoll);
      console.log('editPoll--->',editPoll.questions[questionIndex]);
      let questionId = editPoll.questions[questionIndex].id;
      if(questionId){
        this.props.deletePollQuestionAction(questionId)
      }
      editPoll.questions.splice(questionIndex, 1)
      return { editPoll };
    });
  }
  render() {
    console.log("edit render", this.state);
    let { editPoll } = this.state;
    let startDate = moment
      .unix(editPoll.start_date)
      .format("YYYY-MM-DD HH:mm:ss");
    console.log(
      "0000000000000000",
      moment.unix(editPoll.start_date).format("YYYY-MM-DD HH:mm:ss"),
      startDate
    );
    return (
      <div style={{ width: "60%" }}>
        {/* Title */}
        <label htmlFor="title" className="title">
          Title
        </label>
        <input
          id="title"
          name="title"
          type="text"
          className="input-title"
          value={editPoll.title}
          onChange={this.onChangeHandledTitle}
        />

        {/* description */}
        <label htmlFor="description" className="desc-title">
          Description
        </label>
        <textarea
          id="description"
          name="description"
          type="text"
          className="text-area-input input-title"
          value={editPoll.description}
          onChange={this.onChangeHandledTitle}
        />
        {editPoll.questions &&
          editPoll.questions.map((question, index) => {
            return (
              <>
                <div className="questions" key={question.id}>
                  <label htmlFor="title" className="title">{`Question ${index +
                    1}`}</label>
                  <div className="dispaly-flex">
                    <input
                      id={`question_${index}`}
                      name={`question_${question.id}`}
                      type="text"
                      className="input-question input-title"
                      value={question.question}
                      index={index}
                      question_id={question.id}
                      onChange={e =>
                        this.onChangeHandledQuestion(
                          e.target.value,
                          index,
                          question.id
                        )
                      }
                    />
                    <span style={{ padding: "0.5em" }}>
                      <MinusCircleOutlined
                        style={{
                          padding: "10px 5px"
                        }}
                        onClick={() => this.removeQuestion(index)}
                      />
                    </span>
                  </div>
                  {question.options &&
                    question.options.map((option, optIndex) => {
                      return (
                        <div className="options" key={option.id}>
                          <div className="dispaly-flex">
                            <input
                              id={`option_${option.id}`}
                              name={`option_${option.id}`}
                              type="text"
                              className="input-question input-title"
                              value={option.option}
                              onChange={e =>
                                this.onChangeHandledOption(
                                  e.target.value,
                                  index,
                                  optIndex
                                )
                              }
                            />
                            <span style={{ padding: "0.5em" }}>
                              <MinusCircleOutlined
                                style={{
                                  padding: "10px 5px"
                                }}
                                onClick={() => this.removeOption(index, optIndex)}
                              />
                            </span>
                          </div>
                        </div>
                      );
                    })}
                </div>
                <div className="options">
                  <Button
                    type="dashed"
                    size="small"
                    onClick={() => this.addOption(index)}
                  >
                    <PlusOutlined />
                    Add Option
                  </Button>
                </div>
                <div className="margin10">
                  <label htmlFor="description" className="desc-title">
                    Type
                  </label>
                  <Radio.Group
                    defaultValue={question.type ? 1 : 0}
                    size="small"
                    buttonStyle="solid"
                    onChange={e => this.onChangeType(e.target.value, index)}
                  >
                    <Radio.Button value={0}>Sinle</Radio.Button>
                    <Radio.Button value={1}>Multiple</Radio.Button>
                  </Radio.Group>
                </div>
              </>
            );
          })}
        <div className="questions">
          <Button type="dashed" size="small" onClick={() => this.addQuestion()}>
            <PlusOutlined />
            Add Question
          </Button>
        </div>
        <Row>
          <Col span={8}>
            <label htmlFor="description" className="desc-title">
              From
            </label>
            <DatePicker
              style={{
                width: "90%",
                padding: "5px 10px",
                margin: "8px 0",
                boxSizing: "border-box"
              }}
              showTime
              format="YYYY-MM-DD HH:mm:ss"
              //defaultValue={moment(startDate, 'YYYY-MM-DD HH:mm:ss')}
            />
          </Col>
          <Col span={8}>
            <label htmlFor="description" className="desc-title">
              To
            </label>
            <DatePicker
              style={{
                width: "90%",
                padding: "5px 10px",
                margin: "8px 0",
                boxSizing: "border-box"
              }}
              showTime
              format="YYYY-MM-DD HH:mm:ss"
              //defaultValue={moment('2015-01-01', 'YYYY-MM-DD')}
            />
          </Col>
        </Row>
        <Radio.Group
          defaultValue={editPoll.status == "draft" ? "draft" : "published"}
          size="small"
          buttonStyle="solid"
          onChange={e => this.onChangeStatus(e.target.value)}
        >
          <Radio value="draft">Save as draft</Radio>
          <Radio value="published">Publish</Radio>
        </Radio.Group>
        <div className="btn-update">
          <Button type="primary" onClick={this.updatePoll}>
            update Poll
          </Button>
        </div>
      </div>
    );
  }
}
const mapStateToProps = state => {
  console.log("~~~~~~~~~~~~~edit-polls state--------", state);
  return {
    editPoll: state.dashboard.editPoll,
    editDetails: state.dashboard.editDetails
  };
};

const mapDispatchToProps = {
  getPollAction: getPoll,
  editPollAction: editPoll,
  deletePollOptionAction: deletePollOption,
  deletePollQuestionAction: deletePollQuestion
};
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditPoll);
