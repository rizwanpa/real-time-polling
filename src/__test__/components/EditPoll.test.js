import React from "react";
import { Provider } from "react-redux";
import { shallow, configure } from "enzyme";
import toJson from "enzyme-to-json";
import Adapter from "enzyme-adapter-react-16";
import configureStore from "redux-mock-store"; // Smart components

// Component to be tested
import EditPoll from "./../../components/EditPoll";
configure({ adapter: new Adapter() });
// create any initial state needed
const initialState = {
  polls: [],
  pollsAnalytics: [],
  deletedPollId: "",
  editPoll : [],
  editDetails : {}
};
// here it is possible to pass in any middleware if needed into //configureStore
const mockStore = configureStore();
const store = mockStore(initialState);

describe("EditPoll Component", () => {
  let wrapperEditPoll;
  // our mock EditPoll function to replace the one provided by mapDispatchToProps
  const mockEditPollfn = jest.fn();
  describe("render()", () => {
    test("renders the component", () => {
      const wrapper = shallow(
        <Provider store={store}>
          <EditPoll />
        </Provider>
      );
      const component = wrapper.dive();

      expect(toJson(component)).toMatchSnapshot();
    });
  });
});
