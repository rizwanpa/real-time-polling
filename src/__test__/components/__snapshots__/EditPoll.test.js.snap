// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditPoll Component render() renders the component 1`] = `<Connect(EditPoll) />`;
